### PR TITLE
Backpopulate stable ScoreV2 scores with ScoreV2 system mod (and don't recalculate their total score)

### LINKED ISSUE #24122: Scores which were set with osu!stable "Score V2" mod convert incorrectly to standardised
Since https://github.com/ppy/osu/pull/24072, we are now converting all legacy scores to new standardised. Unfortunately, we are not accounting for imported scores which were originally set with the "Score V2" mod. Such scores are currently getting double-converted.

The proposed direction is to read the mod value and store it as a system/legacy mod, then skip conversion (ie. rollback to `LegacyTotalScore`).

### PR DESCRIPTION
**Pull request by bdach**
Closes https://github.com/ppy/osu/issues/24122.

As per [agreed](https://discord.com/channels/188630481301012481/188630652340404224/1126001536552685632) [direction](https://discord.com/channels/188630481301012481/188630652340404224/1125754013003038800), I'm not doing anything fancy to hide the mod from the users. Wording of description is up for debate.

I don't currently have a large collection of stable scores to test on. I wanted to test on [this one](https://discord.com/channels/188630481301012481/188630652340404224/1125691508859871253) but it's no longer available. While I _did_ test this, and exercised the edge cases I could think of, it was only done on a single score and in a semi-artificial manner. Please test judiciously.

### CODE REVIEW

File: osu.Game/Database/RealmAccess.cs [hunk 1]
```diff
@@ -986,6 +991,64 @@ private void applyMigrationsForVersion(Migration migration, ulong targetVersion)
...
+
+                                    // we should have this in LegacyTotalScore already, but if we're reading through this anyways...
+                                    int totalScore = sr.ReadInt32();
+
+                                    sr.ReadUInt16(); // Max combo.
+                                    sr.ReadBoolean(); // Perfect.
+
+                                    var legacyMods = (LegacyMods)sr.ReadInt32();
+
+                                    if (!legacyMods.HasFlagFast(LegacyMods.ScoreV2) || score.APIMods.Any(mod => mod.Acronym == @"SV2"))
```
Comment (bdach) [inline osu.Game/Database/RealmAccess.cs:1036]: The second condition here should be unreachable, but I felt safer with it.

File: osu.Game/Database/RealmAccess.cs [hunk 2]
```diff
@@ -986,6 +991,64 @@ private void applyMigrationsForVersion(Migration migration, ulong targetVersion)
...
+                                        continue;
+
+                                    score.APIMods = score.APIMods.Append(new APIMod(new ModScoreV2())).ToArray();
```
Comment (bdach) [inline osu.Game/Database/RealmAccess.cs:1039]: `APIMods` is used here for speed. `Mods` is [orders of magnitude slower](https://gist.github.com/bdach/29f4e84525a7b40274540b2eff80e29c) due to relying on reflection.

**Comment by WitherFlower** [comment]
ScoreV2 on stable causes the multipliers of some mods to change, for example HR goes from 1.06 to 1.10. 
If the goal is to keep those scores, they should probably be adjusted to the right lazer multipliers.

**Comment by bdach** [comment]
Yeah I'm not sure what to do about that.

```csharp
        internal static double ScoreMultiplierV2(Mods modStatus, PlayModes mode, Beatmap beatmap)
        {
            double multiplier = 1;
            switch (mode)
            {
                case PlayModes.Osu:
                    if (CheckActive(modStatus, Mods.HardRock))
                    {
                        modStatus &= ~Mods.HardRock;
                        multiplier *= 1.10;
                    }


                    if (CheckActive(modStatus, Mods.DoubleTime))
                    {
                        modStatus &= ~Mods.DoubleTime;
                        multiplier *= 1.20;
                    }
                    break;
                case PlayModes.CatchTheBeat:
                    if (CheckActive(modStatus, Mods.Hidden))
                        modStatus &= ~Mods.Hidden;
                    break;
            }


            if (CheckActive(modStatus, Mods.NoFail))
                modStatus &= ~Mods.NoFail;


            return multiplier * ScoreMultiplier(modStatus, mode, beatmap);
        }
```

I guess undoing those wouldn't be _very_ hard but it wasn't discussed in any way, shape, or form up to now.

**Comment by smoogipoo** [comment]
We're going to have to rethink/re-adjust the multipliers because that can't be a thing.

**Comment by bdach** [comment]
> Probably end up using the SV2 multipliers there over the existing ones.

That sounds like a separate PR to me (and potentially even changes to lazer mod multipliers) - agree?

File: osu.Game/Database/RealmAccess.cs [hunk 3]
```diff
@@ -986,6 +991,64 @@ private void applyMigrationsForVersion(Migration migration, ulong targetVersion)
...
+                        if (!score.IsLegacyScore || !score.Ruleset.IsLegacyRuleset())
+                            continue;
+
+                        string? replayFilename = score.Files.FirstOrDefault(f => f.Filename.EndsWith(@".osr", StringComparison.InvariantCultureIgnoreCase))?.File.GetStoragePath();
+                        if (replayFilename == null)
+                            continue;
+
+                        try
+                        {
+                            using (var stream = files.Store.GetStream(replayFilename))
```
Comment (peppy) [inline osu.Game/Database/RealmAccess.cs:1008]: Can we move as much of this as possible to a helper method inside `StandardisedScoreMigrationTools`?

Comment (bdach) [reply osu.Game/Database/RealmAccess.cs]: Does https://github.com/ppy/osu/pull/24169/commits/9e4ffc8c12cdce5bb34fee06fafdec76ac41337e work for you?

Comment (peppy) [reply osu.Game/Database/RealmAccess.cs]: Seems a bit better yeah.

**Comment by peppy** [comment]
Tested with the previous breaking cases, looking good now.

osu! 2023-07-11 at 08 04 02


@smoogipoo will leave this approved in case you have anything to add.

**Review by peppy — APPROVED** [review]

**Comment by smoogipoo** [comment]
> That sounds like a separate PR to me (and potentially even changes to lazer mod multipliers) - agree?

Agree

**Pull request merged**